### PR TITLE
Add munit clues tutorial

### PR DESCRIPTION
**Pull request by rochala**
image

### CODE REVIEW

File: _overviews/toolkit/munit-clues.md [hunk 1]
```diff
@@ -9,8 +9,135 @@ next-page: munit-asynchronous-tests
...
+# values are not the same
+# => Obtained
+# JohnSmith
+# => Diff (- obtained, + expected)
+# -JohnSmith
+# +John Smith
+```
+
+Sometimes we may want to change it to better suit our needs.
+In MUnit there are two ways to achieve it. First one allows us to directly provide
```
Comment (julienrf) [inline _overviews/toolkit/munit-clues.md:54]:
```suggestion
In MUnit, there are two ways to achieve it. The first one allows us to directly provide
```

File: _overviews/toolkit/munit-clues.md [hunk 2]
```diff
@@ -9,8 +9,135 @@ next-page: munit-asynchronous-tests
...
+custom message to assertion functions as an additional parameter, for example:
```
Comment (julienrf) [inline _overviews/toolkit/munit-clues.md:55]:
```suggestion
a custom message to assertion functions as an additional parameter, for example:
```

File: _overviews/toolkit/munit-clues.md [hunk 3]
```diff
@@ -9,8 +9,135 @@ next-page: munit-asynchronous-tests
...
+# MyTests:
+# ==> X MyTests.price after discount can't be lower than 0  munit.FailException: 
+#   ./example/MyTests.test.scala:7 price cannot have negative value: -5.0
+# 9:       s"price cannot have negative value: ${price - discount}"
+# 10:    )
+# 11:  }
+```
+
+The other way to change output messages is to use `clue` method for those arguments which values we want to be printed
+in case of failure:
```
Comment (julienrf) [inline _overviews/toolkit/munit-clues.md:100]: This sentence is not very clear, could you maybe reword it or break it into multiple sentences? It could be useful to show an example without using `clue`, involving a complex expression, to show that the default report does not help to focus on the important parts of the expression, hence the need for `clue`.

File: _overviews/toolkit/munit-clues.md [hunk 4]
```diff
@@ -9,8 +9,135 @@ next-page: munit-asynchronous-tests
 
 {% include markdown.html path="_markdown/install-munit.md" %}
 
-<!-- 
-Tutorial about clues.
 
-See https://scalameta.org/munit/docs/assertions.html#assert
- -->
\ No newline at end of file
+In MUnit tests outputs are presented to the user in different ways depending on used assertion.
```
Comment (bishabosha) [inline _overviews/toolkit/munit-clues.md:13]:
```suggestion
In MUnit tests outputs are presented to the user in different ways depending on which assertions are used.
```

File: _overviews/toolkit/munit-clues.md [hunk 5]
```diff
@@ -9,8 +9,135 @@ next-page: munit-asynchronous-tests
...
+For example `assertEquals` failure returns not only stack trace, 
+but also by default includes reason of the failure: `values are not the same` along with differences between two values.
```
Comment (bishabosha) [inline _overviews/toolkit/munit-clues.md:15]:
```suggestion
but also by default includes the reason of the failure: `values are not the same` along with differences between the two values.
```

Comment (adpi2) [inline _overviews/toolkit/munit-clues.md:15]: The first sentences should be a summary of the tutorial.

Something along those lines:
```
In MUnit we can use `clues` to get better error reports. There are two kind of clues:
- the `clue` argument of the `assert` method to print an explanation of the error
- the `clue` method to choose which values to print in the error report
```

Then we can have two parts.

## Using the `clue` argument

## Using the `clue` method

File: _overviews/toolkit/munit-clues.md [hunk 6]
```diff
@@ -9,8 +9,135 @@ next-page: munit-asynchronous-tests
...
+
+{% tabs munit-assert-equals %}
+{% tab 'Scala CLI' %}
```
[the remark this answers is not in the document]
Comment (adpi2) [reply _overviews/toolkit/munit-clues.md:18]: But also, the tutorial should be self-contained. One user should not need to read more than one tutorial to solve its problem.

So I think it is actually better if the first code-block contains a full example. But then the subsequent examples can only contain the `assert`.

File: _overviews/toolkit/munit-clues.md [hunk 7]
```diff
@@ -9,8 +9,135 @@ next-page: munit-asynchronous-tests
...
+
+class MyTests extends munit.FunSuite:
+  test("price after discount can't be lower than 0") {
+    val price = 10.0
+    val discount = 15.0
+    assert(
+      price - discount >= 0,
+      s"price cannot have negative value: ${price - discount}"
+    )
+  }
```
Comment (adpi2) [inline _overviews/toolkit/munit-clues.md:71]: This should be the first code block.

**Review by adpi2 — COMMENTED** [review]
Thanks @rochala for writing this tutorial. At the very beginning it was not clear to me what kind of tutorials we should write. But thanks to the Spree, and also thanks to Szymon's feedback, I now have a better idea of how to write them.

In short, we would like the tutorials to be concise and give the most useful information straightforwardly.

In particular:
- the first sentence should be a summary of the tutorial
- the first code block of each part should contain the solution

**Comment by szymon-rd** [comment]
I think a better title would be just "How to get better failing test messages". 
I would also skip the first snippet and just describe it like: 
> Normally, the MUnit will only provide the `values are not the same` error message.

File: _overviews/toolkit/munit-clues.md (hunk 5 above)
Comment (szymon-rd) [reply _overviews/toolkit/munit-clues.md:15]: Optionally, there could also be a short introduction to the problem, to get the user to feel that we are speaking exactly about the problem that he came to us with. Like
```
Normally, the MUnit will only provide the `values are not the same` error message. It's often not enough for making readable test reports and we can use `clues` to get better error reports. There are two kind of clues:
- the `clue` argument of the `assert` method to print an explanation of the error
- the `clue` method to choose which values to print in the error report
```

**Comment by SethTisue** [comment]
@adpi were thinking maybe this is a niche topic could be relegated to the "What else can MUnit do?" section, with a very short example and a link to MUnit's own documentation on the subject.

(And if MUnit's own documentation on the subject is poor, perhaps this PR could be adapted and merged there instead.)
